### PR TITLE
add update case ref handler

### PR DESCRIPTION
**Pull request by mdayican**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-766

### Change description ###

add handler method to make call to pay-hub

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java [hunk 1]
```diff
@@ -54,4 +57,28 @@ public CreatePaymentResponse handlePaymentMessage(PaymentMessage paymentMessage)
...
+            "Sending payment update case reference request, envelope id {}, case ref {}, exception ref {}",
+            paymentMessage.envelopeId,
+            request.ccdCaseNumber,
+            paymentMessage.exceptionRecordRef
+        );
+
+        ResponseEntity<?> response = payHubClient.updateCaseReference(
+            authTokenGenerator.generate(),
+            paymentMessage.exceptionRecordRef,
+            request
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: could be inlined

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java [hunk 2]
```diff
@@ -54,4 +57,28 @@ public CreatePaymentResponse handlePaymentMessage(PaymentMessage paymentMessage)
...
+        );
+
+        log.info("Payment update response from PayHub, envelope id {}, http status {} ",
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: worth adding `:` in the message for better readability imho.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java [hunk 3]
```diff
@@ -54,4 +57,28 @@ public CreatePaymentResponse handlePaymentMessage(PaymentMessage paymentMessage)
...
+                 paymentMessage.envelopeId,
+                 response.getStatusCode());
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: formatting

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java [hunk 4]
```diff
@@ -60,4 +73,60 @@ public void should_call_payhub_api_for_successful_payment_message() {
...
+
+        // when
+        messageHandler.updatePaymentCaseReference(message);
+
+        // then
+        ArgumentCaptor<CaseReferenceRequest> requestCaptor = ArgumentCaptor.forClass(CaseReferenceRequest.class);
+
+        verify(payHubClient).updateCaseReference(any(), any(), requestCaptor.capture());
+        CaseReferenceRequest req = requestCaptor.getValue();
+        CaseReferenceRequest request = new CaseReferenceRequest("cas-ref-9999");
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: perhaps `expectedRequest`?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java [hunk 5]
```diff
@@ -60,4 +73,60 @@ public void should_call_payhub_api_for_successful_payment_message() {
...
+        assertThat(req).isEqualToComparingFieldByFieldRecursively(request);
+    }
+
+
+    @Test
+    public void should_throw_exception_payhub_if_assign_case_ref_throws() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: something wrong here

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java [hunk 6]
```diff
@@ -47,7 +55,12 @@ public void should_call_payhub_api_for_successful_payment_message() {
         PaymentMessage message = SamplePaymentMessageData.paymentMessage("1234", true);
 
         when(s2sTokenGenerator.generate()).thenReturn("test-service");
-        CreatePaymentRequest request = new CreatePaymentRequest("1234", singletonList("1234"), true, "test-siteId");
+        CreatePaymentRequest request = new CreatePaymentRequest(
+            "1234",
+            singletonList("1234"),
+            true,
+            "test-siteId"
+        );
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: ?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java (hunk 1 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: not understand

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java (hunk 2 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: added

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java (hunk 3 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: formatted

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java (hunk 6 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: revert

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java (hunk 4 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: changed the name

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java [hunk 7]
```diff
@@ -60,4 +68,60 @@ public void should_call_payhub_api_for_successful_payment_message() {
...
+        ).thenThrow(new PayHubClientException(new HttpClientErrorException(HttpStatus.BAD_REQUEST, "BAD_REQUEST")));
+
+        // when
+        assertThatThrownBy(
+            () -> messageHandler.updatePaymentCaseReference(message))
+            .isInstanceOf(PayHubClientException.class);
+
+        // then
+        verify(payHubClient).updateCaseReference(any(), any(), any());
+    }
```
Comment (aliveni) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: Is there a validation on the fields in the message?

Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: this test only tests the exception . the test above  this test verify that.

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: kind of pointless `verify` step as it is stubbed with "exact" values few lines above

**Comment by mdayican** [comment]
pipeline stuck. 
release "bulk-scan-payment-processor-pr-62" does not exist. Installing it now.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java (hunk 7 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageHandlerTest.java]: removed, it is in new PR #64

**Pull request closed without merging**